### PR TITLE
Modularize api

### PR DESCRIPTION
**Pull request by bhavyaagg**
I have modularized the API routes.
But i think there are some parts which can be improved:

1) 
HTTP verbs should not be present in the URL
https://github.com/scorelab/Stackle/blob/master/stackle_api/app/routes.js#L148

HTTP verb "delete" is also used in the URL

2)
Singular and Plural both are used for an object in the URL
i.e. 
We have used /posts as well as /post. 
https://github.com/scorelab/Stackle/blob/master/stackle_api/app/routes.js#L25
https://github.com/scorelab/Stackle/blob/master/stackle_api/app/routes.js#L49

It should be /api/posts and /api/posts/:postid respectively

Same for orgs and org

3)
https://github.com/scorelab/Stackle/blob/master/stackle_api/app/routes.js#L160
This can be /api/stack/subscribe

These are just my opinions. I know I might be wrong. But would love to work on Stackle :)

### CODE REVIEW

**Comment by bhavyaagg** [comment]
Should we use single quote(') or double quote(") in the project?

**Comment by psnmissaka** [comment]
Good job @bhavyaagg Modular
1. I agree with not having HTTP verbs in the API endpoint URL, it's if better we can update the API end points like that going forward.
2. The reason of having both singular and plural object names was to differentiate API calls to a single post and for all posts. 
3. Yes. We have to change it. end point naming will be consistent.

We are planning on following Airbnb style guide for Stackle. So it's better to always use single quotes and backticks as necessary. It's also better to use ES6 syntax as well.

**Comment by bhavyaagg** [comment]
@psnmissaka Thank you for reviewing my PR. Sorry I am not well today. Will update my PRs tomorrow morning. I think we can use eslint also. Would add that too. 

For the plural case I think we can do /posts/:postid for single post i guess?

**Comment by agentmilindu** [comment]
@bhavyaagg I agree on both 1 and 2! I'm not sure about 3, but it sounds good.

**Comment by agentmilindu** [comment]
@padamchopra @iammosespaulr Please review.

**Comment by psnmissaka** [comment]
@bhavyaagg I did some reading and yes what you suggested seems like the better way to do it. Let's follow a resource naming style like `/api/users` for resource collections and `/api/users/:userId` for single resource.

File: stackle_api/app/routes/api_v1/misc.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ * Created by bhavyaagg on 18/02/18.
+ */
+
+
+// The routes present here has to be sorted into meaningful routes
+
+const express = require('express');
+const router = express.Router();
+
+const postModels = require('../../models/post')
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/misc.js]: Add semi-colon at the end of the line

File: stackle_api/app/routes/api_v1/misc.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+
+//delete stack
+router.delete('/delete/stack/:stackid', function (req, res) {
+    let stack_id = req.params.stackid;
+    Stack.remove({_id: stack_id}, function (err, success) {
+        if (err) {
+            res.send("Couldn't delete Stack");
+        } else {
+            res.send("")
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/misc.js]: add semi-colon

File: stackle_api/app/routes/api_v1/misc.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    let query = {_id: postid};
+    let comment = new Comment(req.body);
+    Post.update(query, {comments: []});
+})
+
+
+
+//delete stack
+router.delete('/delete/stack/:stackid', function (req, res) {
+    let stack_id = req.params.stackid;
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/misc.js]: Use Camel Case please

File: stackle_api/server.js [hunk 4]
```diff
@@ -2,44 +2,49 @@ require('dotenv').config();
...
 app.use(bodyParser.json());                                     // parse application/json
-app.use(bodyParser.json({ type: 'application/vnd.api+json' })); // parse application/vnd.api+json as json
+app.use(bodyParser.json({type: 'application/vnd.api+json'})); // parse application/vnd.api+json as json
 app.use(methodOverride());
 
 
-var routes = require('./app/routes')(app,db);
+// var routes = require('./app/routes')(app,db);
+
+app.use('/api', require('./app/routes/api_v1'))
```
Comment (iammosespaulr) [inline stackle_api/server.js]: semi-colon

File: stackle_api/app/routes/misc.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+// The routes present here has to be sorted into meaningful routes
+
+const express = require('express');
+const router = express.Router();
+
+router.get('/home', function (req, res) {
+    //needs to intergrate with github for implementation
+    res.end();
+})
```
Comment (iammosespaulr) [inline stackle_api/app/routes/misc.js]: semi-colon

File: stackle_api/app/routes/api_v1/user.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    let post = new Post(req.body);
+    post.save(function (err, post) {
+        if (err) {
+            console.log("error saving the post");
+            res.send("error!")
+        } else {
+            res.send("Sucessfully saved the post!");
+        }
+    });
+})
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/user.js]: semi-colon

File: stackle_api/app/routes/api_v1/user.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+const Post = postModels.Post;
+
+//save a post
+router.post('/post', function (req, res) {
+    let post = new Post(req.body);
+    post.save(function (err, post) {
+        if (err) {
+            console.log("error saving the post");
+            res.send("error!")
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/user.js]: semi-colon

File: stackle_api/app/routes/api_v1/stack.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+//getting subscribed stacks for a user
+router.get('/subscribed/:userid', function (req, res) {
+    res.header("Access-Control-Allow-Origin", "*");
+    res.header('Access-Control-Allow-Methods: GET, POST, PATCH, PUT, DELETE, OPTIONS');
+    res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
+    User.findOne({userId: req.params.userid}, function (err, result) {
+        if (err) {
+            res.send(err);
+        } else if (result) {
+            let sub_stack = result.subscribed_stacks;
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/stack.js]: Camel Case please

File: stackle_api/app/routes/api_v1/stack.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+            res.send(sub_stack);
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/stack.js]: Camel Case Please

File: stackle_api/app/routes/api_v1/stack.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    stack.save(function (err, stack) {
+        if (err) {
+            console.log("Error saving the stack to database");
+            res.send("Error saving stack!");
+        } else if (stack) {
+            res.send("Sucessfully created the stack");
+        } else {
+            res.send("Null");
+        }
+    })
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/stack.js]: Semi-colon

File: stackle_api/app/routes/api_v1/posts.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+//returns posts relating to specific org
+router.get('/org/:org_name', function (req, res) {
+    res.header("Access-Control-Allow-Origin", "*");
+    res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
+    let orgname = req.params.org_name;
+    Post.find({org_name: orgname}, function (err, posts) {
+        if (err) {
+            console.log(`Error getting posts from $orgname`);
+        }
+        else
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/posts.js]: expected { after else

File: stackle_api/app/routes/api_v1/posts.js [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        res.send(posts);
+    })
+})
+
+//returns posts relating to specific org
+router.get('/org/:org_name', function (req, res) {
+    res.header("Access-Control-Allow-Origin", "*");
+    res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
+    let orgname = req.params.org_name;
+    Post.find({org_name: orgname}, function (err, posts) {
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/posts.js]: Camel case please

File: stackle_api/app/routes/api_v1/posts.js [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+            console.log("Erorr getting posts");
+        res.send(posts);
+    })
+})
+
+//returns posts relating to specific org
+router.get('/org/:org_name', function (req, res) {
+    res.header("Access-Control-Allow-Origin", "*");
+    res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
+    let orgname = req.params.org_name;
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/posts.js]: Camel case please

File: stackle_api/app/routes/api_v1/posts.js [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    let id = req.params.user;
+    Post.find({user: id}, function (err, posts) {
+        if (err)
+            console.log("Erorr getting posts");
+        res.send(posts);
+    })
+})
+
+//returns posts relating to specific org
+router.get('/org/:org_name', function (req, res) {
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/posts.js]: Camel case please

File: stackle_api/app/routes/api_v1/posts.js [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+
+//get all posts
+router.get('/', function (req, res) {
+    res.header("Access-Control-Allow-Origin", "*");
+    res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
+    Post.find({}, function (err, posts) {
+        if (err)
+            console.log("Cant get all posts!")
+        res.send(posts);
+    })
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/posts.js]: semi-colon

File: stackle_api/app/routes/api_v1/posts.js [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+const Post = postModels.Post;
+
+//get all posts
+router.get('/', function (req, res) {
+    res.header("Access-Control-Allow-Origin", "*");
+    res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
+    Post.find({}, function (err, posts) {
+        if (err)
+            console.log("Cant get all posts!")
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/posts.js]: semi-colon

File: stackle_api/app/routes/api_v1/posts.js [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+const postModels = require('../../models/post')
+
+const Post = postModels.Post;
+
+//get all posts
+router.get('/', function (req, res) {
+    res.header("Access-Control-Allow-Origin", "*");
+    res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
+    Post.find({}, function (err, posts) {
+        if (err)
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/posts.js]: expected { after if

File: stackle_api/app/routes/api_v1/posts.js [hunk 18]
```diff
@@ -0,0 +1,47 @@
+/**
+ * Created by bhavyaagg on 17/02/18.
+ */
+
+const express = require('express');
+const router = express.Router();
+
+const postModels = require('../../models/post')
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/posts.js]: semi-colon

File: stackle_api/app/routes/api_v1/post.js [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+    res.header("Access-Control-Allow-Origin", "*");
+    res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
+    let objectid = req.params.postid;
+    Post.findOne({_id: objectid}, function (err, post) {
+        if (err) {
+            res.send(err)
+        } else {
+            res.send(post);
+        }
+    })
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/post.js]: semi-colon

File: stackle_api/app/routes/api_v1/post.js [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+const Post = postModels.Post;
+
+//get a post by id
+router.get('/:postid', function (req, res) {
+    res.header("Access-Control-Allow-Origin", "*");
+    res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
+    let objectid = req.params.postid;
+    Post.findOne({_id: objectid}, function (err, post) {
+        if (err) {
+            res.send(err)
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/post.js]: semi-colon

File: stackle_api/app/routes/api_v1/post.js [hunk 21]
```diff
@@ -0,0 +1,40 @@
+/**
+ * Created by bhavyaagg on 17/02/18.
+ */
+const express = require('express');
+const router = express.Router();
+
+const postModels = require('../../models/post')
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/post.js]: semi-colon

File: stackle_api/app/routes/api_v1/orgs.js [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+const Stack = require('../../models/stack');
+
+//get all stacks (orgs)
+router.get('/', function (req, res) {
+    Stack.find({}, function (err, stacks) {
+        if (err)
+            console.log("Errors retrieving stacks!");
+        else
+            res.send(stacks);
+    })
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/orgs.js]: semi-colon

File: stackle_api/app/routes/api_v1/orgs.js [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+const router = express.Router();
+
+const Stack = require('../../models/stack');
+
+//get all stacks (orgs)
+router.get('/', function (req, res) {
+    Stack.find({}, function (err, stacks) {
+        if (err)
+            console.log("Errors retrieving stacks!");
+        else
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/orgs.js]: expected { after else

File: stackle_api/app/routes/api_v1/orgs.js [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+
+const express = require('express');
+const router = express.Router();
+
+const Stack = require('../../models/stack');
+
+//get all stacks (orgs)
+router.get('/', function (req, res) {
+    Stack.find({}, function (err, stacks) {
+        if (err)
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/orgs.js]: expected { after if

File: stackle_api/app/routes/api_v1/org.js [hunk 25]
```diff
@@ -0,0 +1,24 @@
...
+    res.header("Access-Control-Allow-Origin", "*");
+    res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
+    let orgname = req.params.orgname;
+    Stack.find({name: orgname}, function (err, org) {
+        if (err) {
+            console.log('Error');
+        } else {
+            res.send(org);
+        }
+    })
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/org.js]: semi-colon

File: stackle_api/app/routes/api_v1/misc.js [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+        }
+    })
+})
+
+//user subscribing to an stack
+router.post('/subscribe', function (req, res) {
+    let userid = req.body.uid;
+    let stackname = req.body.stack_name;
+    let query = {userId: userid};
+    User.findOneAndUpdate(query, {$push: {subscribed_stacks: stackname}}, function (err, noaffected) {
```
Comment (iammosespaulr) [inline stackle_api/app/routes/api_v1/misc.js]: camel case

**Comment by iammosespaulr** [comment]
Also we have this one ```
Conflicting files
stackle_api/server.js
```

**Comment by bhavyaagg** [comment]
I have updated my PR

File: stackle_api/server.js [hunk 27]
```diff
@@ -15,8 +16,12 @@ app.use(bodyParser.json());                                     // parse applica
 app.use(bodyParser.json({type: 'application/vnd.api+json'})); // parse application/vnd.api+json as json
 app.use(methodOverride());
 
-// No need to store it into a variable
-require("./app/routes")(app, db);
+// var routes = require('./app/routes')(app,db);
+
+app.use('/api', require('./app/routes/api_v1'))
```
[the remark this answers is not in the document]
Comment (bhavyaagg) [reply stackle_api/server.js]: Sorry I missed them. Updating.

**Comment by bhavyaagg** [comment]
@padamchopra I have updated the PR

File: stackle_api/app/routes/api_v1/misc.js (hunk 26 above)
Comment (bhavyaagg) [reply stackle_api/app/routes/api_v1/misc.js]: Will have to change the whole schema. Will update the models in a different PR as a lot of changes would be there.

**Comment by bhavyaagg** [comment]
I have updated the PR to incorporate the necessary changes

**Comment by bhavyaagg** [comment]
@psnmissaka @padamchopra @agentmilindu Can this be merged?

**Comment by bhavyaagg** [comment]
@psnmissaka @agentmilindu I have tested the API Endpoints as well.

**Pull request closed without merging**